compressors/tANS.py

```diff
@@ -0,0 +1,380 @@
+        """
+        # calculate the power of 2 lying in [freq[s], 2freq[s] - 1]
+        y = get_bit_width(self.params.max_shrunk_state[s])
+        num_out_bits_base = self.params.NUM_STATE_BITS - y
+
+        # calculate the threshold to output 1 more bit
+        thresh_state = (self.params.max_shrunk_state[s] + 1) << num_out_bits_base
+        return num_out_bits_base, thresh_state
+
+    def build_base_encode_step_table(self):
+        """
+        cached version of rans_encoder.rans_base_encode_step function
+        """
+        rans_encoder = rANSEncoder(self.params)
+        self.base_encode_step_table = {}  # M rows, each storing x_next in [L,H]
+        for s in self.params.freqs.alphabet:
+            _min, _max = self.params.min_shrunk_state[s], self.params.max_shrunk_state[s]
+            for x_shrunk in range(_min, _max + 1):
+                self.base_encode_step_table[(s, x_shrunk)] = rans_encoder.rans_base_encode_step(
+                    s, x_shrunk
+                )
+
+    def build_shrink_num_out_bits_lookup_table(self):
+        """
+        caching the shrink_state_num_out_bits_base function
+        """
+        self.shrink_state_num_out_bits_base_table = {}
+        self.shrink_state_thresh_table = {}
+        for s in self.params.freqs.alphabet:
+            num_bits, thresh = self.shrink_state_num_out_bits_base(s)
+            self.shrink_state_num_out_bits_base_table[s] = num_bits
+            self.shrink_state_thresh_table[s] = thresh
+
+    def _print_lookup_tables(self):
+        """
+        function useful to visualize the tANS tables + debugging
+        """
+        print("-" * 20)
+        print("base encode step table")
+        pprint.pprint(self.base_encode_step_table)
+        print("-" * 20)
+        print("num out bits table")
+        pprint.pprint(self.shrink_state_num_out_bits_base_table)
+        print("-" * 20)
+        print("shrink state thresh")
+        pprint.pprint(self.shrink_state_thresh_table)
+
+    def encode_symbol(self, s, state: int) -> Tuple[int, BitArray]:
+        """Encodes the next symbol, returns some bits and the updated state
+
+        In the tANS encode_symbol, note that all we are doing is accessing lookup tables.
+        The lookup tables are already defined during the init
+
+        Args:
+            s (Any): next symbol to be encoded
+            state (int): the rANS state
+
+        Returns:
+            state (int), symbol_bitarray (BitArray):
+        """
+        # output bits to the stream so that the state is in the acceptable range
+        # [L, H] *after*the `rans_base_encode_step`
+        symbol_bitarray = BitArray("")
+
+        # shrink state x before calling base encode
+        num_out_bits = self.shrink_state_num_out_bits_base_table[s]
+        if state >= self.shrink_state_thresh_table[s]:
+            num_out_bits += 1
+
+        out_bits = uint_to_bitarray(state)[-num_out_bits:] if num_out_bits else BitArray("")
```

Comment:
basic python question: why does this work? num_out_bits is not binary? I tried locally for an int with 0 but it throws error for me.